sstables/kl/reader_impl.hh

```diff
@@ -190,224 +169,159 @@ public:
-            }
-            break;
-        }
-        case state::RANGE_TOMBSTONE:
-            if (read_short_length_bytes(data, _val) != read_status::ready) {
-                _state = state::RANGE_TOMBSTONE_2;
-                break;
-            }
-        case state::RANGE_TOMBSTONE_2:
-            if (read_32(data) != read_status::ready) {
-                _state = state::RANGE_TOMBSTONE_3;
-                break;
-            }
-        case state::RANGE_TOMBSTONE_3:
-            if (read_64(data) != read_status::ready) {
-                _state = state::RANGE_TOMBSTONE_4;
-                break;
+                _sst->get_stats().on_row_read();
+                auto ret = _consumer.consume_row_start(key_view(to_bytes_view(_key)), del);
+                // after calling the consume function, we can release the
+                // buffers we held for it.
+                _key.release();
+                _state = state::ATOM_START;
+                if (ret == row_consumer::proceed::no) {
+                    co_yield row_consumer::proceed::no;
+                }
             }
-        case state::RANGE_TOMBSTONE_4:
-        {
-            deletion_time del;
-            del.local_deletion_time = _u32;
-            del.marked_for_delete_at = _u64;
-            auto ret = _shadowable
-                     ? _consumer.consume_shadowable_row_tombstone(to_bytes_view(_key), del)
-                     : _consumer.consume_range_tombstone(to_bytes_view(_key), to_bytes_view(_val), del);
-            _key.release();
-            _val.release();
-            _state = state::ATOM_START;
-            if (ret == row_consumer::proceed::no) {
-                return row_consumer::proceed::no;
+            skip_row = false;
+            while(true) {
+                if (read_short_length_bytes(*_processing_data, _key) != read_status::ready) {
+                    co_yield row_consumer::proceed::yes;
+                }
+                if (_u16 == 0) {
+                    // end of row marker
+                    _state = state::ROW_START;
+                    co_yield _consumer.consume_row_end();
+                    break;
+                }
+                if (read_8(*_processing_data) != read_status::ready) {
+                    co_yield row_consumer::proceed::yes;
+                }
+                _state = state::NOT_CLOSING;
+                auto const mask = column_mask(_u8);
+
+                if ((mask & (column_mask::range_tombstone | column_mask::shadowable)) != column_mask::none) {
+                    _shadowable = (mask & column_mask::shadowable) != column_mask::none;
+                    if (read_short_length_bytes(*_processing_data, _val) != read_status::ready) {
+                        co_yield row_consumer::proceed::yes;
+                    }
+                    if (read_32(*_processing_data) != read_status::ready) {
+                        co_yield row_consumer::proceed::yes;
+                    }
```

Comment:
Perhaps read_32 (and friends) and return a value, for which co_yield takes the following actions:
 - if the value indicates we can continue, the awaiter does nothing (await_suspend returns false)
 - if the value indicates we can't continue but should continue pushing data, does what it currently does for `yes`
 - similary for no

So we can write `co_yield read_32(*_processing_data);`

Reply:
Nice idea, though to avoid the cost of suspending the coroutine, it will probably be ```await_ready``` instead of ```await_suspend``` (like in ```suspend_never```)

Reply:
Yes.